art/attacks/evasion/auto_projected_gradient_descent.py

```diff
@@ -115,11 +115,30 @@ def __init__(
                     if is_probability(estimator.predict(x=np.ones(shape=(1, *estimator.input_shape)))):
                         raise NotImplementedError("Cross-entropy loss is not implemented for probability output.")
 
-                    self._loss_object = tf.reduce_mean(
-                        tf.keras.losses.categorical_crossentropy(
-                            y_pred=estimator._output, y_true=estimator._labels_ph, from_logits=True
-                        )
-                    )
+                    # self._loss_object = tf.reduce_mean(
+                    #     tf.keras.losses.categorical_crossentropy(
+                    #         y_pred=estimator._output, y_true=estimator._labels_ph, from_logits=True
+                    #     )
+                    # )
+                    # modification for image-wise stepsize update
+                    class CrossEntropyLoss:
+                        def __init__(self, reduction="mean"):
+                            self.ce = tf.keras.losses.categorical_crossentropy(
+                                y_pred=estimator._output, y_true=estimator._labels_ph, from_logits=True
+                            )
+                            self.reduction = reduction
+
+                        def __call__(self, y_true, y_pred):
+                            if self.reduction == "mean":
+                                return tf.reduce_mean(self.ce(y_true, y_pred))
+                            elif self.reduction == "sum":
+                                return tf.reduce_sum(self.ce(y_true, y_pred))
+                            elif self.reduction == "none":
+                                return self.ce(y_true, y_pred)
+                            else:
+                                raise NotImplementedError()
+
+                    self._loss_object = CrossEntropyLoss()
```

Comment:
It looks like the tests for ACG are failing for TensorFlow v1. I think it could be related to the way the class `CrossEntropyLoss` is used this line 141. I think this creates an instance of `CrossEntropyLoss` but does not return the values of `__call__`.

Reply:
Thank you for your comment. I think the following code will work. Is it correct?
```python
class CrossEntropyLoss:
    def __init__(self, reduction="mean"):
        self.reduction = reduction

    def __call__(self, y_true, y_pred):
        ce_loss = tf.keras.losses.categorical_crossentropy(
            y_pred=y_pred, y_true=y_true, from_logits=True
        )
        if self.reduction == "mean":
            return tf.reduce_mean(ce_loss)
        elif self.reduction == "sum":
            return tf.reduce_sum(ce_loss)
        elif self.reduction == "none":
            return ce_loss
        else:
            raise NotImplementedError()

self._loss_object = CrossEntropyLoss()(y_pred=estimator._output, y_true=estimator._labels_ph)
```

Reply:
I think this should almost work. For some reason the code scanning raised a warning `Non-callable called` (https://github.com/Trusted-AI/adversarial-robustness-toolbox/security/code-scanning/383) for this version. Maybe we can avoid the warning by first creating an instance:
```python
loss = CrossEntropyLoss()
self._loss_object = loss(y_pred=estimator._output, y_true=estimator._labels_ph)
```